components/service/glean/src/main/java/mozilla/components/service/glean/storages/GenericScalarStorageEngine.kt

```diff
@@ -0,0 +1,203 @@
+            deserializeSingleMetric(metricValue)?.let {
+                storeData[metricName] = it
+            } ?: logger.warn("Failed to deserialize $metricStoragePath")
+        }
+
+        return prefs
+    }
+
+    /**
+     * Retrieves the [recorded metric data][ScalarType] for the provided
+     * store name.
+     *
+     * Please note that the [Lifetime.Application] lifetime is handled implicitly
+     * by never clearing its data. It will naturally clear out when restarting the
+     * application.
+     *
+     * @param storeName the name of the desired store
+     * @param clearStore whether or not to clear the requested store. Not that only
+     *        metrics stored with a lifetime of [Lifetime.Ping] will be cleared.
+     *
+     * @return the [ScalarType] recorded in the requested store
+     */
+    @Synchronized
+    fun getSnapshot(storeName: String, clearStore: Boolean): GenericDataStorage<ScalarType>? {
+        val allLifetimes: GenericDataStorage<ScalarType> = mutableMapOf()
+
+        // Make sure data with "user" lifetime is loaded before getting the snapshot.
+        // We still need to catch exceptions here, as `getAll()` might throw.
+        @Suppress("TooGenericExceptionCaught")
+        try {
+            userLifetimeStorage.all
+        } catch (e: NullPointerException) {
+            // Intentionally left blank. We just want to fall through.
+        }
+
+        // Get the metrics for all the supported lifetimes.
+        for (store in dataStores) {
+            store[storeName]?.let {
+                allLifetimes.putAll(it)
+            }
+        }
+
+        if (clearStore) {
+            // We only allow clearing metrics with the "ping" lifetime.
+            dataStores[Lifetime.Ping.ordinal].remove(storeName)
+        }
+
+        return if (allLifetimes.isNotEmpty()) allLifetimes else null
+    }
+
+    /**
+     * Get a snapshot of the stored data as a JSON object.
+     *
+     * @param storeName the name of the desired store
+     * @param clearStore whether or not to clearStore the requested store
+     *
+     * @return the [JSONObject] containing the recorded data.
+     */
+    override fun getSnapshotAsJSON(storeName: String, clearStore: Boolean): Any? {
+        return getSnapshot(storeName, clearStore)?.let { dataMap ->
+            return JSONObject(dataMap)
+        }
+    }
+
+    /**
+     * Helper function for the derived classes. It can be used to record
+     * simple scalars to the internal storage.
+     */
+    protected fun recordScalar(
+        stores: List<String>,
+        category: String,
+        name: String,
+        lifetime: Lifetime,
+        value: ScalarType
+    ) {
+        checkNotNull(applicationContext) { "No recording can take place without an application context" }
+
+        // Record a copy of the metric in all the needed stores.
+        @SuppressLint("CommitPrefEdits")
```

Comment:
Why is suppressing the lint needed? Is it not smart enough to see that `apply` is called at the end if it is non-null?

Reply:
No, it's not :( The IDE will keep telling me `apply` is missing without that suppression.